### PR TITLE
Don't consider NaN ordered against any number

### LINKED ISSUE #106: Matchers handle double.nan unintuitively and inconsistently
Example 1:
```dart
expect(double.nan, greaterThan(10)); // pass
expect(double.nan, lessThan(-10)); // fail
```
In my opinion, both of these should fail. I can't imagine a situation where someone would write a test like `expect(myFunc(), greaterThan(10));` and then actually expect the test to pass if `myFunc()` returns nan. The intuitive behavior is that such a test fails.

Example 2:
```dart
expect(double.nan, lessThanOrEqualTo(0)); // fail
expect(double.nan, inInclusiveRange(double.negativeInfinity, 0)); // pass
```
This is inconsistent. Mathematically speaking, aren't `x <= 0` and ` -inf <= x <= 0` equivalent? But here the tests behave differently.

Version: matcher 0.12.3+1

### PR DESCRIPTION
**Pull request by natebosch**
Fixes #106

### CODE REVIEW

**Comment by natebosch** [comment]
We'll need to be careful rolling this out - hopefully no code is relying on the broken behavior...

**Comment by natebosch** [comment]
cc @jonahwilliams @DaveShuckerow @Hixie - this breaks some Flutter tests.

For instance there are some tests which are comparing `Offset` instances with `greaterThan`:
https://github.com/flutter/flutter/blob/2c558fdb213ecf9ed0dd47b9231c3ad9f33061e0/packages/flutter/test/material/reorderable_list_test.dart#L925

However `Offset` is not guaranteed to have a total ordering, and so some of these tests cases change behavior when we account for that in the matcher.

https://github.com/flutter/engine/blob/8427d73c8340c9e51492689ee66261a4a72cac9e/lib/ui/geometry.dart#L44-L45


Can you weigh in on how you'd like to move forward? Are you OK with a bug fix that breaks some usage like this? Will you have any difficulty fixing these tests?

File: lib/src/numeric_matchers.dart [hunk 1]
```diff
@@ -73,7 +73,8 @@ class _InRange extends FeatureMatcher<num> {
     if (value == _high) {
       return _highMatchValue;
     }
-    return true;
+    // Value may still be outside if range if it can't be compared
```
Comment (grouma) [inline lib/src/numeric_matchers.dart]: [nit] period.

File: lib/src/numeric_matchers.dart [hunk 2]
```diff
@@ -63,7 +63,7 @@ class _InRange extends FeatureMatcher<num> {
   const _InRange(
       this._low, this._high, this._lowMatchValue, this._highMatchValue);
 
-  bool typedMatches(value, Map matchState) {
+  bool typedMatches(num value, Map matchState) {
```
Comment (grouma) [inline lib/src/numeric_matchers.dart]: Can we change the signature on the `typedMatches` method above as well? Also is this not a breaking change or is this not exposed directly?

Comment (natebosch) [reply lib/src/numeric_matchers.dart]: oh, I think I shouldn't have included this. I'm pretty sure we're getting it explicitly.

**Comment by jonahwilliams** [comment]
Sorry this one flew under the radar for me. Look at the effect on the flutter repo we have two failing tests. Both of these seem to be test assertions which are broken as currently written:

```
Expected: a value greater than Offset:<Offset(400.0, 528.0)>
Actual: Offset:<Offset(400.0, 480.0)>
```

```
 Expected: a value greater than ?:<Size(800.0, 34.0)>
 Actual: ?:<Size(800.0, 102.0)>
```

if this change will help us write better tests then I'm all for it. For the process of updating flutter to use the latest matcher, I can coordinate the changes there.

Also, with this change I did notice that we've been implementing `>=` incorrectly for Offset and Size since basically forever (https://github.com/flutter/engine/pull/8210) - so that is definitely something.

**Comment by Hixie** [comment]
@jakemac53 
> That being said this is technically a breaking change, but I will leave that up to you.

Technically _everything_ is a breaking change. It's easy for someone to depend on the exact commit hash in their code, for example, in which case changing the commit hash will break them.

**Pull request merged**